divi/src/BlockFactory.cpp

```diff
@@ -96,7 +95,10 @@ CBlockTemplate* BlockFactory::CreateNewBlock(const CScript& scriptPubKeyIn, bool
         block.nVersion = settings.GetArg("-blockversion", block.nVersion);
     }
 
-    pblocktemplate->previousBlockIndex = chain_.Tip();
+    {
+        LOCK(mainCS_);
```

Comment:
We may be able to drop the lock altogether. I was doing some digging into it, and since the chain tip is already captured and passed as a parameter downstream, the lock isnt really required. Moreover, since we're checking the chainTip for being non-null, it would function correctly. Do you have any objections to removing this lock altogether?

Reply:
I think strictly speaking the lock is required, since `CChain::Tip()` performs some operations on `std::vector` (like `operator[]`).  And those are likely not thread-safe, e.g. the vector could just be in the process of being reallocated or something.  My guess is that in practice it will most likely work almost always, but in theory it could lead to weird bugs.

It might be possible to make `CChain::Tip()` thread safe by caching the tip pointer and using e.g. `std::atomic` inside `CChain`, which is something we could do at some point if you are worried about the lock.  But in the current state, I don't think it would be (completely) safe to remove the lock.

In the end, this is your decision, though.  As I said, I don't think it would cause a lot of issues in practice; so if you think we should remove the lock, there are no objections from my side.